examples/fluids/problems/stats.c

```diff
@@ -0,0 +1,558 @@
+  CeedMemType        ceed_mem_type;
+  PetscFunctionBeginUser;
+
+  PetscLogStage stage_stats_process;
+  PetscCall(PetscLogStageGetId("Stats Process", &stage_stats_process));
+  if (stage_stats_process == -1) PetscCall(PetscLogStageRegister("Stats Process", &stage_stats_process));
+  PetscCall(PetscLogStagePush(stage_stats_process));
+
+  CeedGetPreferredMemType(user->ceed, &ceed_mem_type);
+  CeedVectorSetValue(user_stats.parent_stats, 0);
+
+  CeedVectorGetArrayRead(user_stats.child_stats, ceed_mem_type, &child_stats);
+  CeedVectorGetArray(user_stats.parent_stats, ceed_mem_type, &parent_stats);
+
+  if (user_stats.num_comp_stats == 1) unit = MPIU_REAL;
+  else {
+    PetscCallMPI(MPI_Type_contiguous(user_stats.num_comp_stats, MPIU_REAL, &unit));
+    PetscCallMPI(MPI_Type_commit(&unit));
+  }
+
+  PetscCall(PetscSFReduceBegin(user_stats.sf, unit, child_stats, parent_stats, MPI_SUM));
+  PetscCall(PetscSFReduceEnd(user_stats.sf, unit, child_stats, parent_stats, MPI_SUM));
+
+  CeedVectorRestoreArrayRead(user_stats.child_stats, &child_stats);
+  CeedVectorRestoreArray(user_stats.parent_stats, &parent_stats);
+
+  CeedVectorScale(user_stats.parent_stats, 1 / user_stats.span_width);
+
+  // L^2 projection with the parent_data
+  PetscCall(DMGetGlobalVector(user_stats.dm, &rhs));
+  PetscCall(DMGetLocalVector(user_stats.dm, &rhs_loc));
+  PetscCall(VecZeroEntries(rhs));
+  PetscCall(VecZeroEntries(rhs_loc));
+  PetscCall(VecGetArrayWriteAndMemType(rhs_loc, &rhs_arr, &rhs_mem_type));
+  CeedVectorSetArray(user_stats.rhs_ceed, MemTypeP2C(rhs_mem_type), CEED_USE_POINTER, (PetscScalar *)rhs_arr);
+
+  CeedOperatorApply(user_stats.op_stats_proj, user_stats.parent_stats, user_stats.rhs_ceed, CEED_REQUEST_IMMEDIATE);
+
+  CeedVectorTakeArray(user_stats.rhs_ceed, MemTypeP2C(rhs_mem_type), &rhs_arr);
+  PetscCall(VecRestoreArrayAndMemType(rhs_loc, &rhs_arr));
+  PetscCall(DMLocalToGlobal(user_stats.dm, rhs_loc, ADD_VALUES, rhs));
+
+  PetscCall(VecDuplicate(rhs, stats));
+  PetscCall(VecPointwiseMult(*stats, rhs, user_stats.M_inv));
+
+  PetscCall(KSPSolve(user_stats.ksp, rhs, *stats));
+
+  PetscCall(PetscLogStagePop());
+  PetscFunctionReturn(0);
+}
+
+// TSMonitor for the statistics collection and processing
+PetscErrorCode TSMonitor_Statistics(TS ts, PetscInt steps, PetscReal solution_time, Vec Q, void *ctx) {
+  User user = (User)ctx;
+  Vec  stats;
+  PetscFunctionBeginUser;
+
+  // Do not collect or process on the first step of the run (ie. on the initial condition)
+  if (steps == user->app_ctx->cont_steps) PetscFunctionReturn(0);
+
+  if (steps % user->app_ctx->stats_collect_interval == 0) {
+    PetscCall(CollectStatistics(user, solution_time, Q));
+  }
+
+  if (steps % user->app_ctx->stats_write_interval == 0 && user->app_ctx->stats_write_interval != -1) {
+    PetscCall(DMGetGlobalVector(user->sp_stats.dm, &stats));
+    PetscCall(ProcessStatistics(user, &stats));
+    PetscCall(VecViewFromOptions(stats, NULL, "-stats_write_view"));
```

Comment:
Better to hold the `PetscViewer` in the monitor so this can be a `VecView(stats, viewer)`.

Reply:
Done